exporter.py

```diff
@@ -21,7 +23,113 @@
+
+    # set logging credentials
+    global logger
+    logger = set_logging_args(args.log_file, args.log_level)
+
+    # parse config file if provided as an argument
+    if args.config_file:
+        args = parseConfig(args)
+
+    # Get username and password of Citrix ADC
+    ns_user, ns_password = get_login_credentials(args)
+
+    # Wait for other containers to start.
+    logger.info('Sleeping for %s seconds.' % args.start_delay)
+    time.sleep(args.start_delay)
+
+    # Load the metrics file specifying stats to be collected
+    metrics_json = get_metrics_file_data(args.metrics_file, args.metric)
+
+    # Get protocol type to access ADC
+    ns_protocol = get_ns_session_protocol(args) 
+
+    # Get cert validation args provided 
+    ns_cert = get_cert_validation_args(args, ns_protocol) 
+
+    requests.packages.urllib3.disable_warnings(InsecureRequestWarning)
+    requests.packages.urllib3.disable_warnings(SubjectAltNameWarning)
+
+    # Creates the global session with default timeout
+    ns_session = newSession(timeout=args.timeout)
+
+    # Start the server to expose the metrics.
+    start_exporter_server(args.port)
+
+    if not args.k8sCICprefix.isalnum():
+        logger.error('Invalid k8sCICprefix : non-alphanumeric not accepted')
+
+    # Register the exporter as a stat collector
+    logger.info('Registering collector for %s' % args.target_nsip)
+
+    try:
+        REGISTRY.register(CitrixAdcCollector(nsip=args.target_nsip, metrics=metrics_json, username=ns_user,
+                                             password=ns_password, protocol=ns_protocol, 
+                                             k8s_cic_prefix=args.k8sCICprefix, ns_cert=ns_cert,
+                                             ns_session=ns_session))
+    except Exception as e:
+        logger.error('Invalid arguments! could not register collector for {}::{}'.format(args.target_nsip, e))
+
+    # Forever
+    while True:
+        signal.pause()
+
+
+class TimeoutHTTPAdapter(HTTPAdapter):
+    def __init__(self, *args, **kwargs):
+        if "timeout" in kwargs:
+            self.timeout = kwargs["timeout"]
+            del kwargs["timeout"]
+        if "max_retries" in kwargs:
+            self.max_retries = kwargs["max_retries"]
+            del kwargs["max_retries"]
+        super().__init__(*args, **kwargs)
+
+    def get(self, request, **kwargs):
+        timeout = kwargs.get("timeout")
+        if timeout is None:
+            kwargs["timeout"] = self.timeout
+        return super().get(request, **kwargs)
+    
+    def post(self, request, **kwargs):
+        timeout = kwargs.get("timeout")
+        if timeout is None:
+            kwargs["timeout"] = self.timeout
+        return super().post(request, **kwargs)
+
+def newSession(retries=3, backoff_factor=1, timeout=None, session=None):
+    session = session or requests.Session()
+    timeout = timeout or 15
+    retry = Retry(
+            total=retries,
+            backoff_factor=backoff_factor
+    )
```

Comment:
From what i read, retry by default won't work on POST as it results to multiple inserts. Did you try the same. It worked ?

Reply:
i have some tests here and will try it out as soon as i am able, thanks for reviewing the code! :)

Reply:
After due consideration, we are not in favour of exiting the exporter if certain retries fails. Best approach is to ignore the subsequent prometheus requests till ADC comes up or session resumes with a well defined backoff strategy. Will come up with that soon. 

Meanwhile, can you specify some detail on other issue that you mentioned, wherein you wrote:
"i'm still not sure how to handle parcial failures though (e.g. can get some stats, but not all)."

What kind of partial failures? is it any particular stats that you were not able to fetch?

Reply:
there are some metrics that can be disabled in adc and a warning is replied - that's what i meant with partial failure - and that's ok... kind of too verbose to log that on every request tough.

yeah, exiting is not the solution. best approach recommended by prometheus authors is to keep the exporter up answering the up metric as 0 (no such metric as of now) - and keep trying to reach adc in a gentle manner.

i will try to migrate this code to fit ha_proxy exporter written by the prometheus authors and see how it goes.

Reply:
Regarding those warnings, we are changing the loglevels for such messages. Most of these will be moved to DEBUG while default being INFO. So, going forwards, these logs won't be an issue. One can always enable DEBUG as per need. 
For other changes, yes, there won't be any exit(apart from invalid username/password, as it can't be corrected anyways), will incorporate in next release.